frontend/src/metabase/containers/SaveQuestionModal.tsx

```diff
@@ -0,0 +1,220 @@
+import FormCollectionPicker from "metabase/collections/containers/FormCollectionPicker/FormCollectionPicker";
+import Form from "metabase/core/components/Form";
+import FormInput from "metabase/core/components/FormInput";
+import FormFooter from "metabase/core/components/FormFooter";
+import FormTextArea from "metabase/core/components/FormTextArea";
+import FormErrorMessage from "metabase/core/components/FormErrorMessage";
+import Button from "metabase/core/components/Button";
+import FormSubmitButton from "metabase/core/components/FormSubmitButton";
+import FormRadio from "metabase/core/components/FormRadio";
+import { canonicalCollectionId } from "metabase/collections/utils";
+import { CollectionId } from "metabase-types/api";
+import * as Errors from "metabase/core/utils/errors";
+import Question from "metabase-lib/Question";
+
+import "./SaveQuestionModal.css";
+
+const getSingleStepTitle = (questionType: string, showSaveType: boolean) => {
+  if (questionType === "model") {
+    return t`Save model`;
+  } else if (showSaveType) {
+    return t`Save question`;
+  }
+
+  return t`Save new question`;
+};
+
+const SAVE_QUESTION_SCHEMA = Yup.object({
+  saveType: Yup.string(),
+  name: Yup.string().when("saveType", {
+    // We don't care if the form is valid when overwrite mode is enabled,
+    // as original question's data will be submitted instead of the form values
+    is: (value: string) => value !== "overwrite",
+    then: Yup.string().required(Errors.required),
+    otherwise: Yup.string().nullable(true),
+  }),
+});
+
+interface SaveQuestionModalProps {
+  question: Question;
+  originalQuestion: Question | null;
+  onCreate: (question: Question) => void;
+  onSave: (question: Question) => Promise<void>;
+  onClose: () => void;
+  multiStep?: boolean;
+  initialCollectionId?: number;
+}
+
+interface FormValues {
+  saveType: string;
+  collection_id: CollectionId | null | undefined;
+  name: string;
+  description: string;
+}
+
+const isOriginalQuestionNotNullable = (
+  question: Question | null,
+  saveType: string,
+): question is Question => {
+  return saveType === "overwrite";
+};
+
+export default class SaveQuestionModal extends Component<SaveQuestionModalProps> {
+  handleSubmit = async (details: FormValues) => {
+    const { question, originalQuestion, onCreate, onSave } = this.props;
+
+    const collectionId = canonicalCollectionId(
+      isOriginalQuestionNotNullable(originalQuestion, details.saveType)
+        ? originalQuestion.collectionId()
+        : details.collection_id,
+    );
+    const displayName = isOriginalQuestionNotNullable(
+      originalQuestion,
+      details.saveType,
+    )
+      ? originalQuestion.displayName()
+      : details.name.trim();
+    const description = isOriginalQuestionNotNullable(
+      originalQuestion,
+      details.saveType,
+    )
+      ? originalQuestion.description()
+      : details.description
+      ? details.description.trim()
+      : null;
```

Comment:
Is it possible to refactor this so we can call `isOriginalQuestionNotNullable` only once?

Reply:
yes, possible, I thought having the previous structure would simplify review

Reply:
reworked this, two separate methods seems more readable here